### DISCUSSION TITLE
Решение для "переполненных" мест на карте

### DISCUSSION THREAD

**Discussion by paul-k-pastvu** [PastVu/pastvu#462] [Ideas / Идеи и предложения]
Предлагаю обсудить здесь возможные _технические_ пути решения  проблемы "переполненных" фотографиями точек на карте (типа Красной площади в Москве и других подобных мест) . Пример, актуальный для большинства крупных и мелких городов б. СССР - однотипные фотографии демонстраций.

Когда проектировался сайт, эта задача, вероятно, считалась не самой приоритетной. Однако за минувшие годы она назрела, созрела, и, кажется, даже перезрела. 

Никакими орг.вопросами (заменять/не заменять, удалять/оставлять варианты, что публиковать, что отклонять) эта проблема полностью не решается. Но для той же Красной площади, даже если удалить все дубли и варианты, число снимков на квадратный метр всё равно окажется чрезмерно большим. И оно всё время будет расти. Попытки сдерживать этот напор, просто не публикуя все фото, которые кажутся не несущими ничего нового, дает только временную передышку, и только в одном месте. 

Нужно именно техническое решение, которое позволит всем относительно спокойно в этой ситуации существовать. 

Моя идея такова:
1. Ввести новый статус изображений - "дополнительное". 
2. В момент публикации модератор может выбрать, публиковать изображение как "основное" или как "дополнительное". 
3. Модератор в любой момент может переводить изображения из статуса "основное" в статус "дополнительное" и обратно. Это отражается в истории снимка.
4. Текущий статус "опубликовано" по-прежнему остается основным статусом публичных изображений, и присваивается по умолчанию.
5. На карту добавить новый переключатель - "показать дополнительные изображения" (или короче, - "показать все"). 
5.1. По умолчанию он выключен, при этом дополнительные изображения на карте не отображаются - по умолчанию отображается "основной" слой. 
5.2. При включенной опции "показать все", кроме "основных" на карте отображаются _ещё_ и точки "дополнительных" изображений. Опции "показ _только_ дополнительных"  я не предполагал - считаю нецелесообразным.
5.3. Переключатель активен (доступен для выбора), только если на отображаемой части карты есть хотя бы одно изображение в статусе "дополнительного". 
5.4. Этот же переключатель добавляется в фильтр галерей. 

6. Под мини-картой, где сейчас мини-галерея "Ближайшие изображения", добавляется ещё одна галерея "Дополнительные изображения"

Если у кого-то есть _другие_ идеи, предлагайте их прямо здесь, или на ГитХабе - где кому удобнее. Через неделю примерно я вынесу какой-то набор вариантов уже для публичного обсуждения. 

----
Дополнения:
7. не только "жесткое" изменение статуса модератором, но и через инструмент "рейтинга" (лайки/дизлайки, причем с разным "весом", в зависимости о статуса и стажа/вклада участника).

**Comment by paul-k-pastvu** [comment]
Эта идея, естественно, не отменяет и на заменяет необходимости замены версий #75 , но может как-то пересекаться с задачей поддержки "вариантов" (ранее не специфицировалась).

Также (косвенно) м.б. связанно с задачей - #164 (Дополнительная типизация изображений)

**Comment by paul-k-pastvu** [comment]
@anvaruk предлагает "просто увеличить зум". Подход, подкупающий своей простотой. 
Но дело в том, что все эти точки на карте в таких местах в основном _не_нужны_. А нужны они только по особому запросу.
Кроме того, увеличив зум, мы сужаем видимую на экране область. И интересные пользователю снимки с гораздо большей вероятностью окажутся "за кадром". Обычно удается решать свои задачи, "гоняя" зум туда/обратно. Но не в таких местах.

То есть я обеими руками "за" увеличить зум ;) Но это не решение.

**Comment by paul-k-pastvu** [comment]
Ещё одноо дополнение от @anvaruk - не только "жесткое" изменение статуса модератором, но и через инструмент "рейтинга" (лайки/дизлайки, причем с разным "весом", в зависимости о статуса и стажа/вклада участника).

**Comment by kabalin** [comment]
Я бы предложил ограничивать количество точек на карте:

1. Ставим лимит на количество показанных снимков на единицу площади карты (подобрать так, чтобы количество не затрудняло навигацию).
2. Если в базе количество изображений превышает лимит - появляется уведомление о том, что показаны не все снимки из имеющихся для данного фрагмента карты, предлагающее пользователю уточнить временной период или увеличить зум для показа всех изображений. Уведомление убирается в момент когда количество снимков с учетом фильтра становится ниже лимита.

В дополнение, как предлагает @anvaruk - включаем зум до 22 (с незначительным размытием плиток карты ввиду интерполяции).

**Comment by paul-k-pastvu** [comment]
Это идет вразрез с применяемой сейчас технологией "объединения в кластеры".
По факту, это не позволит использовать никакие зумы кроме максимальных.

Повторюсь, проблема не в зуме, а в том, что в принципе далеко не все
снимки, которые есть на сайте, всегда и всем нужно видеть на карте (но и
удалять их тоже нет оснований).
Где-то это не критично, а где-то - уже да.

пн, 14 февр. 2022 г. в 22:59, Ruslan Kabalin ***@***.***>:

> Я бы предложил ограничивать количество точек на карте:
>
>    1. Ставим лимит на количество показанных снимков на единицу площади
>    карты (подобрать так, чтобы количество не затрудняло навигацию).
>    2. Если в базе количество изображений превышает лимит - появляется
>    уведомление о том, что показаны не все снимки из имеющихся для данного
>    фрагмента карты, предлагающее пользователю уточнить временной период для
>    показа всех изображений или увеличить зум. Уведомление убирается в момент
>    когда количество снимков с учетом фильтра становится ниже лимита.
>
> В дополнение, как предлагает @anvaruk <https://github.com/anvaruk> -
> включаем зум до 22 (с незначительным размытием плиток карты ввиду
> интерполяции).

**Reply by kabalin** [reply]
> Это идет вразрез с применяемой сейчас технологией "объединения в кластеры".

Кластеры останутся как есть. Я о лимите показываемых фотографий, который вполне может решить проблему "переполненных" территорий на мой взгляд.

> Повторюсь, проблема не в зуме, а в том, что в принципе далеко не все
снимки, которые есть на сайте, всегда и всем нужно видеть на карте (но и
удалять их тоже нет оснований).

Решение с "дополнительными" снимками действительно на какое-то время уменьшит число показываемых фотографий популярных мест, но, как мне кажется, фундаментально проблему не решит, просто с ростом числа загрузок в критерий "дополнительные" начнут попадать те, что прежде бы считались "основными". 

На самом деле все предложенные решения могут решить проблему большой плотности в той или иной мере. Мы можем исходить из того, как проще ее решить уже сейчас. Зум и лимит точек не сложно добавить. Разделение на "основные" и "дополнительные" - идея хорошая, но реализация потребует больше времени.

**Reply by paul-k-pastvu** [reply]
> > Это идет вразрез с применяемой сейчас технологией "объединения в кластеры".
> Кластеры останутся как есть. Я о лимите показываемых фотографий, который вполне может решить проблему "переполненных" территорий на мой взгляд.

Кто-то из нас не догоняет ;)
Смотри: сейчас, если нельзя отобразить все точки, они "слипаются" в кластеры. Такие (17):
![image](https://user-images.githubusercontent.com/60163100/154062233-c63e2c61-64c7-4adc-bd6c-c746f2419135.png)

Или такие (16):
![image](https://user-images.githubusercontent.com/60163100/154062289-6448d9d5-7fa8-4f70-91c1-98e81d4e61a3.png)

Что будет происходить в случае реализации твоего предложения? В какой именно момент "появляется уведомление о том, что показаны не все снимки из имеющихся для данного фрагмента карты"?

**Reply by kabalin** [reply]
> Что будет происходить в случае реализации твоего предложения? В какой именно момент "появляется уведомление о том, что показаны не все снимки из имеющихся для данного фрагмента карты"?

Когда плотность точек (а возможно и кластеры тоже можно считать за точку на карте, так как они влияют на загруженность карты) будет черезмерная, как например тут:
![image](https://user-images.githubusercontent.com/329780/154136813-f4c72651-95f9-45df-a76f-0179873e99c4.png)
в верхней части карты выводится всплывающее сообщение, о том, что не все точки показаны и предлагающее пользователю уточнить временной интервал (или увеличить зум).

К примеру, пользователь ставит фильтр 1970-1980, новый результат не требует ограничений, навигация более не затруднена:
![image](https://user-images.githubusercontent.com/329780/154137806-3aabc7b7-68f5-4814-985b-7a8c24409ad1.png)

Оценивать плотность стоит от зума 17 и выше (когда в ответе сервера появляются отдельные фото).

На самом деле все немного сложнее, ведь запрос не может гарантировать равномерное распределение расстояния между точками в ответе, поэтому рассчет необходимости лимитирования и скрытие избыточных точек нужно будет делать уже на стороне клиента, когда получен ответ.

**Reply by paul-k-pastvu** [reply]
> К примеру, пользователь ставит фильтр 1970-1980, новый результат не требует ограничений, навигация более не затруднена: !

Мне кажется, что это не решает никакую реальную проблему, при этом создает для пользователя новые искусственные ограничения. Реальная проблема в том, что по умолчанию части точек на карте быть вообще не нужно, ни при каких фильтрах.

В общем, в виде дополнения к моему варианту, это всё годится. Но не как самостоятельное решение.

Частичным решением, кстати, был бы фильтр по направлениям, в дополнение к фильтру по датам -  #240. Вот это было бы мегаудобно.

**Comment by paul-k-pastvu** [comment]
И главное  - ты преложил "ограничивать количество точек на карте". А _какие_, собственного говоря, точки будут показаны-то? Какие именно? По какому принципу движок их отберет?

**Reply by kabalin** [reply]
Я вообще думал случайно выбирать, но можно рассмотреть идеи принципа выборки, если дойдет до реализации.

**Reply by paul-k-pastvu** [reply]
Не, давай лучше не будем в эту сторону двигаться... Нас распнут за "случайную выборку". И за любую другую тоже.

**Comment by kabalin** [comment]
Я посмотрел более внимательно - при больших значениях зума (>=17 aka localWork) кластеры формируются на стороне клиента. Может быть перенастроить функцию так, чтобы большее число точек попадало в кластеры, тем самым уменьшить насыщенность точек для данного значения зума? Критерием может быть любое соприкосновение изображений точек. Чтобы проиллюстрировать, на фрагменте стрелками показано где должен быть кластер (там больше соприкосновений, я не все показал, но идея я думаю понятна).

![image](https://user-images.githubusercontent.com/329780/154170282-4023fa3c-bd4f-448f-aaab-b8702a676685.png)

**Reply by paul-k-pastvu** [reply]
Да, так будет лучше - вместе с добавлением уровней зума до 22 это мягко принудит пользователей увеличивать зум для плотно забитых локаций. Это не решение проблемы как таковой, но улучшение UX. Раз это на клиенте, то, думаю, стоит это просто сделать (в той же задаче, что и расширение шкалы зума).

**Comment by kabalin** [comment]
Отдельный вариант решения проблемы для карты в режиме показа фотографии. Может быть временной диапазон на боковой карте по умолчанию ставить равным текущей датировки фотографии (или диапазону +-N лет от датировки), таким образом число точек значительно снизится. Может даже показаться удобнее - заинтересовал, скажем, меня снимок 50-х годов, я его открыл и все соседние точки на карте тоже отражают этот же период.

**Reply by paul-k-pastvu** [reply]
Согласен. Стоит вынести это в отдельную задачу. Я считаю разумным диапазон +/- 10 лет (т.е. для снимка 1955 года это будет 1945-1965). Но это лучше вынести в отдельную настройку, чтобы пользователь мог сам для себя выбрать этот дефолтный интервал (причем отдельно "назад", и отдельно "вперед").

**Reply by paul-k-pastvu** [reply]
(Ремарка; боковую карту в режиме показа фотографии мы вроде бы решили называть "мини-картой". Предлагаю придерживаться этого термина)
